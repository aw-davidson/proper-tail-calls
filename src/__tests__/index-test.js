const babel = require('babel-core');
const plugin = require('../');

const example = `
function simpleTailCall() {
    return simpleTailCall()
}
`;

const fibExample = `
 function fib (n, last = 1, beforeLast = 0) {
  if (n === 0) {
    return beforeLast
  }

  return fib(n - 1, last + beforeLast, last)
}
`;

const countVowelsExample = `
const isVowel = (char) => /[aeioue]/.test(char.toLowerCase())
function countVowels(count, str) {
  count += isVowel(str[0]) ? 1 : 0;
  if (str.length <= 1) {
    return count;
  }
  return countVowels(count, str.slice(1));
}
`;

it('works', () => {
  const { code } = babel.transform(example, { plugins: [plugin] });
  expect(code).toMatchSnapshot();
});

it('transforms proper tail calls', () => {
  const { code } = babel.transform(fibExample, { plugins: [plugin] });
  const f = new Function(`
    ${code};
    return fib;
  `);
  const fib = f();
  expect(fib(10)).toBe(55);
});

it('does not produce stack overflow where a normal function would', () => {
  const { code } = babel.transform(countVowelsExample, { plugins: [plugin] });
  const f = new Function(`
    ${code};
    return countVowels;
  `);
  const countVowels = f();
  expect(countVowels(0, 'a'.repeat(10000))).toBe(10000);

  const getOverflowFn = new Function(`
  ${countVowelsExample};
  return countVowels;
`);
  const countVowelsOverflow = getOverflowFn();

  try {
    countVowelsOverflow(0, 'a'.repeat(10000));
  } catch (e) {
    expect(e).toEqual(RangeError('Maximum call stack size exceeded'));
  }
});

it('works for ternary expressions', () => {
  const ternaryCode = 'const counter = (n, acc = 0) => n === 0 ? acc : counter(n - 1, acc + 1)';

  const { code } = babel.transform(ternaryCode, { plugins: [plugin] });
  expect(code).toMatchSnapshot();
});
