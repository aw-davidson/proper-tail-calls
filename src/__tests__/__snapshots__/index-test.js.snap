// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it dynamically generates the name of the trampoline function based on the global program scope 1`] = `
"const _trampoline = () => {};

const trampoline = () => {};

function _trampoline2(fn) {
  return function trampolined(...args) {
    let result = fn(...args);

    while (typeof result === 'function') {
      result = result();
    }

    return result;
  };
}"
`;

exports[`works 1`] = `
"var simpleTailCall = _trampoline(function simpleTailCall() {
  return () => {
    return simpleTailCall();
  };
});

function _trampoline(fn) {
  return function trampolined(...args) {
    let result = fn(...args);

    while (typeof result === 'function') {
      result = result();
    }

    return result;
  };
}"
`;

exports[`works for functions with multiple function calls 1`] = `
"var weirdCounter = _trampoline(function weirdCounter(n, acc = 0) {
  if (n <= 0) {
    return acc;
  }

  if (n < 10) {
    return () => {
      return weirdCounter(n - 1, acc + 1);
    };
  }

  if (n < 100) {
    return () => {
      return weirdCounter(n - 1, acc + 10);
    };
  }

  return () => {
    return weirdCounter(n - 1, acc + 100);
  };
});

function _trampoline(fn) {
  return function trampolined(...args) {
    let result = fn(...args);

    while (typeof result === 'function') {
      result = result();
    }

    return result;
  };
}"
`;

exports[`works for ternary expressions with explicit return 1`] = `
"const counter = _trampoline((n, acc = 0) => {
  return () => {
    return n === 0 ? acc : counter(n - 1, acc + 1);
  };
});

function _trampoline(fn) {
  return function trampolined(...args) {
    let result = fn(...args);

    while (typeof result === 'function') {
      result = result();
    }

    return result;
  };
}"
`;

exports[`works for ternary expressions with implicit return 1`] = `
"const counter = _trampoline((n, acc = 0) => () => {
  return n === 0 ? acc : counter(n - 1, acc + 1);
});

function _trampoline(fn) {
  return function trampolined(...args) {
    let result = fn(...args);

    while (typeof result === 'function') {
      result = result();
    }

    return result;
  };
}"
`;
